stream_consumer.go

```diff
@@ -85,57 +90,179 @@ func NewStreamConsumer(streamName string, endpointType EndpointType, endpoints [
 	r.InitialAddrs(addresses)
 	target := r.Scheme() + ":///stream"
 
+	config := defaultStreamEndpointConfig()
+	for _, opt := range opts {
+		opt(config)
+	}
+	conn, err := grpc.Dial(target, grpc.WithInsecure(), grpc.WithBalancerName(roundrobin.Name), grpc.WithInsecure(),
+		grpc.WithDefaultCallOptions(grpc.ForceCodec(&gogoCodec{})),
+		grpc.WithBackoffMaxDelay(config.backoffMaxDelay))
+	if err != nil {
+		return nil, err
+	}
+	endpoint := &StreamEndpoint{
+		config:    config,
+		endpoints: endpoints,
+		target:    target,
+		conn:      conn,
+	}
+
+	return endpoint, nil
+}
+
+func (se *StreamEndpoint) Close() error {
+	return se.conn.Close()
+}
+
+func (se *StreamEndpoint) ConsumeStream(streamName string, opts ...ConsumerConfigOpt) *Consumer {
+
 	config := defaultConsumerConfig()
 	for _, opt := range opts {
 		opt(config)
 	}
 
 	ch := make(chan *stream.Event, config.BufferLen)
-	consumer := &Consumer{
+	c := &Consumer{
 		StreamName: streamName,
 		EvtChan:    ch,
 		config:     config,
-		endpoints:  endpoints,
-		target:     target,
 	}
+
+	var monitoringHolder = consumerMonitoring(streamName, se.endpoints)
+
 	go func() {
-		consumer.run()
+		for se.conn.GetState() != connectivity.Shutdown {
+			waitTillReady(se)
+
+			client := stream.NewStreamClient(se.conn)
+			req := &stream.StreamRequest{Name: streamName}
+
+			var callOpts []grpc.CallOption
+			if config.UseGzip {
+				callOpts = append(callOpts, grpc.UseCompressor(gzip.Name))
+			}
+			callOpts = append(callOpts)
+			st, err := client.Stream(context.Background(), req, callOpts...)
+			if err != nil {
+				Log.Warn("Error while creating stream", zap.String("stream", streamName), zap.Error(err))
+				if se.conn.GetState() == connectivity.Ready {
+					//weird, let's wait before recreating the stream
+					time.Sleep(5 * time.Second)
```

Comment:
should we wait or let GRPC handle this?
Plus, if we have multiple endpoints, shouldn't we let GRPC try the the other endpoint without any sleep?

Reply:
yes, this was meant to wait before recreating the stream if the connection is ready, to avoid a fast spin if the stream cannot be created on a working connection, but this will most likely never happen so it's not worth the extra code